### PR TITLE
Add training suite.

### PR DESCRIPTION
**Pull request by ailzhang**

### CODE REVIEW

File: benchmark/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        """ Returns a function `f` that triggers model training that finishes in 30s.
+        `f` can be called with f(model, example_input)"""
+```
+To test locally before submitting a PR, add the following snippet to `hubconf.py` and run `python hubconf.py`.
+```
+if __name__ == '__main__':
+    m = Model()
+    m.set_device('cpu')
+    model, example_input = m.get_module()
+    m.get_training_loop()(model, example_input)
```
Comment (zdevito) [inline benchmark/README.md:44]: Training loop should probably not take any arguments. Some programs (GANs) won't have a single model. Similarly for the training loop: some things won't train off of the same example inputs that are used by the actual nn.Module. get_training_loop should probably just take a `niterations` argument and then handle its own data loading.

File: benchmark/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    m.get_eval_function()(model, example_input)
```
Comment (zdevito) [inline benchmark/README.md:45]: Similar to above, `model` is not going to be clear for some examples since there might be multiple `nn.Module` objects involved. Though for eval having an example input may be useful.

File: benchmark/utils.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    models = list_models()
+    for model in models:
+        cwd = os.getcwd()
+        model_path = os.path.join(cwd, model_dir, model)
+        sys.path.insert(0, model_path)
+        os.chdir(model_path)
+        hub_module = torch.hub.import_module(hubconf_file, hubconf_file)
+        Model = getattr(hub_module, 'Model', None)
+        if Model:
+            m = Model()
```
Comment (zdevito) [inline benchmark/utils.py:39]: this seems like an extra step since the Model class takes no arguments. Two options:

1. the constructor to Module takes the configuration flags: `jit`, `device`, etc. And we drop the `set_x` functions.
2. We remove the constructor and have `hub_module.model` be an instance of the `Model` class.

I think I like option 1 better since it allows the Model constructor to perform the setup all in one place.

File: benchmark/README.md (hunk 1 above)
Comment (ailzhang) [reply benchmark/README.md:44]: Does `niterations` here refer to number of steps or epochs, or simply just repeat n times? 
The reason I'm asking is not all user code organize functions nicely around step/epoch so I feel repeat n times might be easier?

Comment (ailzhang) [reply benchmark/README.md:44]: Btw if training_loop/eval_function no longer takes any args, shall we just run them when calling them instead of returning the function? 
I'm thinking sth like 
```
m.train(niter=10) # repeat train 10 times
# current
m.get_training_loop()(niter=10)
```

Comment (ailzhang) [reply benchmark/README.md]: I feel `niter` can be a simple thing to do on user end, maybe we could just leave it outside the API for now?

File: benchmark/README.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    def eval(self):
+        """ Optional
+        Runs evaluation on `model` with `example_input` from get_modules()"""
+        raise NotImplementedError()
+```
+To test locally before submitting a PR, add the following snippet to `hubconf.py` and run `python hubconf.py`.
+```
+if __name__ == '__main__':
+    m = Model()
+    m.set_device('cpu')
```
Comment (zdevito) [inline benchmark/README.md:42]: this is still on the old api

File: benchmark/utils.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        _install_deps(model_dir + model)
+
+
+def run_models():
+    models = list_models()
+    for model in models:
+        cwd = os.getcwd()
+        model_path = os.path.join(cwd, model_dir, model)
+        sys.path.insert(0, model_path)
+        os.chdir(model_path)
```
Comment (zdevito) [inline benchmark/utils.py:35]: ideally we would put this logic in the list_models iterator and have `model` be the `Model` class for each backend.

File: benchmark/utils.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                print('Running [{}] on device: {}'.format(model, device))
+                m.get_module()
+                start = time.time()
+                m.train()
+                print('Finished training on device: {} in {}s.'.format(device, time.time() - start))
+                start = time.time()
+                try:
+                    m.eval()
+                except NotImplementedError:
+                    print('Skipped eval since it\'s not implemented for []'.format(model))
```
Comment (zdevito) [inline benchmark/utils.py:50]: `[] -> {}`

File: benchmark/README.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    def get_module(self):
+        """ Required
+        Returns `model`, `example_input`
+        `model` should be torchscript model if self.jit is True.
+        Both `model` and `example_input` should be on `self.device` properly.
+        Both `model` and `example_input` can be lists/dicts as long as it's
+        consumable by train/eval function. """
+
+    def train(self):
```
Comment (zdevito) [inline benchmark/README.md:29]: I think this should take an optional `niterations` which can be used to run a short warmup. It should run niteration passes of the training loop (i.e. forward/backward/optimizerstep). This will allow us to warmup the code by running niterations=1 in benchmarking.

File: benchmark/README.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    def eval(self):
+        """ Optional
+        Runs evaluation on `model` with `example_input` from get_modules()"""
+        raise NotImplementedError()
+```
+To test locally before submitting a PR, add the following snippet to `hubconf.py` and run `python hubconf.py`.
+```
+if __name__ == '__main__':
+    m = Model()
+    m.get_module()
```
Comment (zdevito) [inline benchmark/README.md:42]: we should test that example input works on module:

`module, eg = m.get_module()`
`eg(module)`

**Comment by ailzhang** [comment]
Merging since all comments are resolved. :D
